src/backend/columnar/columnar_customscan.c

```diff
@@ -491,15 +534,430 @@ RelationIdGetNumberOfAttributes(Oid relationId)
 }
 
 
-static Path *
-CreateColumnarScanPath(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
+/*
+ * CheckVarStats() checks whether a qual involving this Var is likely to be
+ * useful based on the correlation stats. If so, or if stats are unavailable,
+ * return true; otherwise return false.
+ */
+static bool
+CheckVarStats(PlannerInfo *root, Var *var, Oid sortop)
+{
+	/*
+	 * Collect isunique, ndistinct, and varCorrelation.
+	 */
+	VariableStatData varStatData;
+	examine_variable(root, (Node *) var, var->varno, &varStatData);
+	if (varStatData.rel == NULL ||
+		!HeapTupleIsValid(varStatData.statsTuple))
+	{
+		return true;
+	}
+
+	AttStatsSlot sslot;
+	if (!get_attstatsslot(&sslot, varStatData.statsTuple,
+						  STATISTIC_KIND_CORRELATION, sortop,
+						  ATTSTATSSLOT_NUMBERS))
+	{
+		ReleaseVariableStats(varStatData);
+		return true;
+	}
+
+	Assert(sslot.nnumbers == 1);
+
+	float4 varCorrelation = sslot.numbers[0];
+
+	ReleaseVariableStats(varStatData);
+
+	/*
+	 * If the Var is not highly correlated, then the chunk's min/max bounds
+	 * will be nearly useless.
+	 */
+	if (Abs(varCorrelation) < ColumnarQualPushdownCorrelation)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * CheckPushdownClause tests to see if clause is a candidate for pushing down
+ * into the given rel (including join clauses). This test may not be exact in
+ * all cases; it's used to reduce the search space for parameterization.
+ *
+ * Note that we don't try to handle cases like "Var + ExtParam = 3". That
+ * would require going through eval_const_expression after parameter binding,
+ * and that doesn't seem worth the effort. Here we just look for "Var op Expr"
+ * or "Expr op Var", where Var references rel and Expr references other rels
+ * (or no rels at all).
+ */
+static bool
+CheckPushdownClause(PlannerInfo *root, RelOptInfo *rel, Expr *clause)
+{
+	if (IsA(clause, OpExpr))
+	{
+		OpExpr *opExpr = (OpExpr *) clause;
+		Expr *lhs = list_nth(opExpr->args, 0);
+		Expr *rhs = list_nth(opExpr->args, 1);
+
+		Var *varSide;
+		Expr *exprSide;
+		if (IsA(lhs, Var) && ((Var *) lhs)->varno == rel->relid)
+		{
+			varSide = castNode(Var, lhs);
+			exprSide = rhs;
+		}
+		else if (IsA(rhs, Var) && ((Var *) rhs)->varno == rel->relid)
+		{
+			varSide = castNode(Var, rhs);
+			exprSide = lhs;
+		}
+		else
+		{
+			return false;
+		}
+
+		/* cannot push down a qual over the whole row */
+		if (varSide->varattno <= 0)
+		{
+			return false;
+		}
+
+		if (contain_volatile_functions((Node *) exprSide))
+		{
+			return false;
+		}
```

Comment:
Maybe we can have a few tests where we explicitly show that we don't push-down filters containing calls to volatile functions.

```sql
EXPLAIN (analyze on, costs off, timing off, summary off)
  SELECT * FROM coltest c1 WHERE c1.x1 = 5090 and (c1.x2 > random() or c1.x3 > 400);
┌─────────────────────────────────────────────────────────────────────────────────┐
│                                   QUERY PLAN                                    │
├─────────────────────────────────────────────────────────────────────────────────┤
│ Custom Scan (ColumnarScan) on coltest c1 (actual rows=1 loops=1)                │
│   Filter: ((x1 = 5090) AND (((x2)::double precision > random()) OR (x3 > 400))) │
│   Rows Removed by Filter: 999                                                   │
│   Columnar Chunk Group Filters: (x1 = 5090)                                     │
│   Columnar Chunk Groups Removed by Filter: 19                                   │
└─────────────────────────────────────────────────────────────────────────────────┘
(5 rows)
```

Reply:
Done.